File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/core/CompanionTicketUrl.kt
// P14 — pure helper that packs a freshly-minted companion-pairing ticket
// into the canonical URL the desktop browser scans:
//
//   https://web.flagshipserver.com/?companion=<base64url(JSON {ticketId,
//   ticketSecret, podBaseUrl, username})>
//
// base64url is RFC 4648 §5 — no padding. The webapp boot-time handler
// reverses this: split off the `companion=` param, base64url-decode,
// JSON.parse, then POST to <podBaseUrl>/api/screens/companion/redeem
// with the (ticketId, ticketSecret) tuple.
//
// MIRRORS: apps/mobile/ios/Sources/FlagshipCore/CompanionTicketUrl.swift
// + apps/web/public/webapp/lib/companion-ticket-url.js.

package com.flagshipserver.app.core

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class CompanionTicketPayload(
    val ticketId: String,
    val ticketSecret: String,
    val podBaseUrl: String,
    val username: String,
)

object CompanionTicketUrl {
    /** Webapp receiver origin, via [Endpoints] (prod-default + test override). */
    private val HOST: String get() = Endpoints.webappOrigin

    private val json = Json {
        explicitNulls = false
        encodeDefaults = true
    }

    fun build(
        ticketId: String,
        ticketSecret: String,
        podBaseUrl: String,
        username: String,
    ): String {
        val payload = CompanionTicketPayload(
            ticketId = ticketId,
            ticketSecret = ticketSecret,
            podBaseUrl = podBaseUrl,
            username = username,
        )
        val jsonString = json.encodeToString(CompanionTicketPayload.serializer(), payload)
        val encoded = Base64URL.encode(jsonString.toByteArray(Charsets.UTF_8))
        return "$HOST?companion=$encoded"
    }
}

data class CompanionDockApprovalPayload(
    val serverDomain: String,
    val requestId: String,
    val approvalSecret: String,
)

object CompanionDockApprovalLink {
    private val hex32 = Regex("^[0-9a-fA-F]{32}$")
    private val hex64 = Regex("^[0-9a-fA-F]{64}$")

    fun parse(raw: String): CompanionDockApprovalPayload? {
        val uri = runCatching { java.net.URI(raw.trim()) }.getOrNull() ?: return null
        if (uri.scheme != "flagship" || uri.host != "dock") return null
        val params = uri.rawQuery.orEmpty().split('&').mapNotNull { pair ->
            val index = pair.indexOf('=')
            if (index < 1) null else {
                val key = java.net.URLDecoder.decode(pair.substring(0, index), "UTF-8")
                val value = java.net.URLDecoder.decode(pair.substring(index + 1), "UTF-8")
                key to value
            }
        }.toMap()
        val server = params["server"]?.lowercase().orEmpty()
        val request = params["request"].orEmpty()
        val code = params["code"].orEmpty()
        if (!server.endsWith(".${Endpoints.dataApex}") || !hex32.matches(request) || !hex64.matches(code)) return null
        return CompanionDockApprovalPayload(server, request.lowercase(), code.lowercase())
    }
}

File: apps/mobile/shared/Sources/FlagshipCore/CompanionTicketURL.swift
import Foundation
import FlagshipAPI

/// P14 — builds the `https://webapp.flagshipserver.com/?companion=...` URL
/// the phone QR encodes. The desktop browser scans the QR, the webapp's
/// boot-time handler decodes the `companion` query param into the
/// 4-tuple `{ ticketId, ticketSecret, podBaseUrl, username }` and POSTs
/// `/api/companion/redeem` against `podBaseUrl`.
///
/// Wire-shape parity matters here: Android + daemon + webapp all decode
/// the same shape. Field names are byte-identical with the daemon's
/// TypeScript `CompanionTicketEnvelope`.
public enum CompanionTicketURL {
    /// Static landing host the webapp lives at. The browser hits this
    /// URL first; the in-page JS parses the `companion` query param and
    /// redirects + redeems against the user's pod. Via `Endpoints`
    /// (prod-default `webapp.flagshipserver.com` + test override).
    public static var webappHost: String { Endpoints.webappHost }

    public struct Envelope: Codable, Equatable, Sendable {
        public let ticketId: String
        public let ticketSecret: String
        public let podBaseUrl: String
        public let username: String

        public init(ticketId: String, ticketSecret: String, podBaseUrl: String, username: String) {
            self.ticketId = ticketId
            self.ticketSecret = ticketSecret
            self.podBaseUrl = podBaseUrl
            self.username = username
        }
    }

    /// Encode the 4-tuple into the canonical
    /// `https://webapp.flagshipserver.com/?companion=<b64url>` URL string.
    /// Returns nil only if the JSON encode itself fails — every other
    /// path is total.
    public static func build(
        ticketId: String,
        ticketSecret: String,
        podBaseUrl: String,
        username: String
    ) -> String? {
        let env = Envelope(
            ticketId: ticketId,
            ticketSecret: ticketSecret,
            podBaseUrl: podBaseUrl,
            username: username
        )
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        guard let json = try? encoder.encode(env) else { return nil }
        let b64 = base64URLNoPadding(json)
        var comps = URLComponents()
        comps.scheme = "https"
        comps.host = webappHost
        comps.path = "/"
        comps.queryItems = [URLQueryItem(name: "companion", value: b64)]
        return comps.url?.absoluteString
    }

    /// Derive `podBaseUrl` from a `PodInfo.fqdn` value the way the
    /// daemon expects: prefix `https://`, no trailing slash. Kept on
    /// this helper so the call site can't drift from the encoder.
    public static func podBaseUrl(forFqdn fqdn: String) -> String {
        let trimmed = fqdn.trimmingCharacters(in: .whitespaces)
        if trimmed.hasPrefix("https://") || trimmed.hasPrefix("http://") {
            return trimmed
        }
        return "https://\(trimmed)"
    }

    /// Base64url-no-padding encoder per RFC 4648 §5.
    static func base64URLNoPadding(_ data: Data) -> String {
        var s = data.base64EncodedString()
        s = s.replacingOccurrences(of: "+", with: "-")
        s = s.replacingOccurrences(of: "/", with: "_")
        while s.hasSuffix("=") { s.removeLast() }
        return s
    }

    /// Companion decoder — re-pads, swaps `-_` back to `+/`, base64-
    /// decodes, then JSON-decodes. Symmetric to `base64URLNoPadding`.
    static func decodeBase64URLNoPadding(_ s: String) -> Data? {
        var t = s
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        while t.count % 4 != 0 { t.append("=") }
        return Data(base64Encoded: t)
    }

    /// Parse a built URL back into its envelope. Used by the round-trip
    /// test + (later) by any in-app handler that wants to inspect a
    /// companion URL it generated. Returns nil if the URL doesn't match
    /// the expected shape.
    public static func parse(_ urlString: String) -> Envelope? {
        guard let url = URL(string: urlString),
              let comps = URLComponents(url: url, resolvingAgainstBaseURL: false),
              let value = comps.queryItems?.first(where: { $0.name == "companion" })?.value,
              let data = decodeBase64URLNoPadding(value),
              let env = try? JSONDecoder().decode(Envelope.self, from: data)
        else { return nil }
        return env
    }
}

public enum CompanionDockApprovalLink {
    public struct Payload: Equatable, Sendable {
        public let serverDomain: String
        public let requestId: String
        public let approvalSecret: String

        public init(serverDomain: String, requestId: String, approvalSecret: String) {
            self.serverDomain = serverDomain
            self.requestId = requestId
            self.approvalSecret = approvalSecret
        }
    }

    /// Accepted deep-link hosts. `remote` is the current label (the feature
    /// was renamed from "dock" on 2026-07-23); `dock` stays accepted so a
    /// browser still serving the pre-rename webapp keeps working against a
    /// freshly built app. The payload is identical either way.
    static let acceptedHosts: Set<String> = ["remote", "dock"]

    public static func parse(_ raw: String) -> Payload? {
        guard let url = URL(string: raw.trimmingCharacters(in: .whitespacesAndNewlines)),
              url.scheme == "flagship",
              let host = url.host, acceptedHosts.contains(host),
              let components = URLComponents(url: url, resolvingAgainstBaseURL: false)
        else { return nil }
        let params = (components.queryItems ?? []).reduce(into: [String: String]()) { out, item in
            if let value = item.value { out[item.name] = value }
        }
        guard let server = params["server"]?.lowercased(),
              server.hasSuffix(".\(Endpoints.dataApex)"),
              let requestId = params["request"]?.lowercased(),
              requestId.range(of: "^[0-9a-f]{32}$", options: .regularExpression) != nil,
              let secret = params["code"]?.lowercased(),
              secret.range(of: "^[0-9a-f]{64}$", options: .regularExpression) != nil
        else { return nil }
        return Payload(serverDomain: server, requestId: requestId, approvalSecret: secret)
    }
}
